integration-healthcheck/src/main/java/org/symphonyoss/integration/healthcheck/AsyncCompositeHealthIndicator.java

```diff
@@ -0,0 +1,185 @@
+   */
+  public void addHealthIndicator(String name, HealthIndicator indicator) {
+    this.indicators.put(name, indicator);
+  }
+
+  @Override
+  public Health health() {
+    try {
+      Map<String, Future<Health>> result = asyncExecution();
+      Map<String, Health> healths = extractResult(result);
+      return this.healthAggregator.aggregate(healths);
+    } catch (InterruptedException e) {
+      String message = "Request thread was interrupted";
+      LOG.error(message, e);
+      return Health.down().withDetail(ERROR_KEY, message).build();
+    }
+  }
+
+  /**
+   * Executes registered indicators using asynchronous calls.
+   * @return
+   * @throws InterruptedException
+   */
+  private Map<String, Future<Health>> asyncExecution() throws InterruptedException {
+    Map<String, Future<Health>> result = new LinkedHashMap<>();
+
+    int indicatorSize = Math.max(1, indicators.size());
+    int poolSize = Math.min(MAX_POOL_SIZE, indicatorSize);
+    ExecutorService service = Executors.newFixedThreadPool(poolSize);
+
+    for (Map.Entry<String, HealthIndicator> entry : indicators.entrySet()) {
+      final HealthIndicator indicator = entry.getValue();
+
+      Future<Health> execution = service.submit(new Callable<Health>() {
+        @Override
+        public Health call() throws Exception {
+          return indicator.health();
+        }
+      });
+
+      result.put(entry.getKey(), execution);
+    }
+
+    shutdownService(service);
+
+    return result;
+  }
+
+  /**
+   * Shutdown the executor service
+   * @param service Executor Service object
+   * @throws InterruptedException Thread execution interrupted
+   */
+  private void shutdownService(ExecutorService service) throws InterruptedException {
+    service.shutdown();
+
+    for (int execution = 0; execution < SHUTDOWN_CHECK_TIMES; execution++) {
+      boolean terminated = service.awaitTermination(EXECUTION_TIMEOUT, TimeUnit.MILLISECONDS);
+
+      if (terminated) {
+        break;
+      }
+    }
+  }
+
+  /**
+   * Extract the result from the asynchronous calls.
+   * @param asyncResult Asynchronous execution result
+   * @return Health indication from all the registered indicators
+   */
+  private Map<String, Health> extractResult(Map<String, Future<Health>> asyncResult) {
+    Map<String, Health> healths = new LinkedHashMap<>();
+
+    for (Map.Entry<String, Future<Health>> entry : asyncResult.entrySet()) {
+      Future<Health> value = entry.getValue();
+
+      Health health = getExecutionValue(value);
+      healths.put(entry.getKey(), health);
+    }
+
+    return healths;
+  }
+
+  /**
+   * Gets the health indication based on the {@link Future} result object.
+   * @param value Asynchronous execution result
+   * @return Health indication
+   */
+  private Health getExecutionValue(Future<Health> value) {
+    try {
+      if (value.isDone()) {
+        return value.get();
+      } else {
+        return Health.unknown().build();
```

Comment:
If you call shutdownNow() above, I don't think you need to check isDone() here.  isDone will return true unless the thread is still running.  It will not be running if interrupted, cancelled or completed normally.